src/language/extension-types.md

```diff
@@ -0,0 +1,315 @@
+
+  var safeId = IdNumber(42424242);
+  safeId + 10; // Compile-time error, no '+' operator.
+  myUnsafeId = safeId; // Compile-time error, wrong type.
+  myUnsafeId = safeId as int; // OK, runtime cast to representation type.
+}
+```
+
+{{site.alert.note}}
+Extension types serve the same purpose as **wrapper classes**,
+but don't require the creation of expensive runtime objects.
+Because extension types are static-only and compiled away at runtime,
+they are essentially zero cost.
+
+[**Extension methods**]() are a static abstraction similar to extension types.
+However, an extension method adds functionality to *all* instances
+of its underlying type. Extension types are the opposite; 
+an extension type's interface only applies to objects
+declared with the extension type,
+and is distinct from the interface of its underlying type by default. 
+Read the [Transparency](#transparency) section for more details.
+{{site.alert.end}}
+
+## Syntax
+
+### Declaration
+
+Define a new extension type with the `extension type` declaration and a name,
+followed by the *representation type declaration* in parenthesis:
+
+```dart
+extension type E(int i) {
+  // Define set of operations
+}
+```
+
+The representation type declaration specifies that 
+the underlying type of extension type `E` is `int`,
+and the reference to the *representation object* is named `i`.
+
+The representation object gives access to the underlying type,
+the same way declaring a final instance variable of the underlying type
+would in a wrapper class.
+It's in scope in the extension type body, and
+you can access it using its name as a getter:
+
+- Within the extension type body: `this.i`.
+- Outside with a property extraction: `e.i` (where `e` is the object of an extension type). 
+
+Extension types declarations can also include [type parameters](generics)
+just like classes:
+
+```dart
+extension type E<T>(int i) {
+  // ...
+}
+```
+
+### Constructors
+
+You can optionally declare constructors in an extension type's body.
+A constructor must be named, because the the representation declaration
+already acts as an unnamed constructor for the extension type
+that initializes the represenation object.
+Any additional constructors must declare the representation object's
+instance variable in its initializer list:
+
+```dart
+// 'E' has a named constructor 'n',
+// as well as its built-in unnamed constructor
+extension type E(int i) {
+  E.n(this.i, String foo);
+}
+
+void main() {
+  var named = E.n(3, "Hello");
+  var unnamed = E(4);
+}
+```
+
+### Members
+
+Declare members in the body of an extension type to define its interface. 
+Extension type member declarations are identical to class members.
+Members can be methods, getters, setters, or operators:
+
+```dart
+extension type NumberE(int value) {
+  // Operator:
+  NumberE operator +(NumberE other) =>
+      NumberE(value + other.value);
+  // Getter:
+  NumberE get i => value;
```

Comment:
Perhaps?:
```suggestion
  NumberE get i => this;
```

We could also return `NumberE(value)`, but `this` is already available, and it is the same object as `value`, and it has the required static type, so why not use it?

Reply:
Sgtm, thanks!